pages/blog/caching-data-with-dataloader.mdx

```diff
@@ -0,0 +1,137 @@
+
+### 2. Create a dataloader instant.
+
+To start, import dataloader and LRU (LRU is a way to map cache data, it is installed with dataloader).
+
+```ts
+import DataLoader from 'dataloader'
+import { LRUMap } from 'lru_map'
+```
+
+### 3. Define the types that would be saved to our cache
+
+The first type should be a way to represent the data, like a string with a name, and the other is the data itself.
+
+```ts
+type PackageName = string
+type PackageInfo = DataType // note that this is the type of the data we represent.
+//for example:
+type DataType = {
+  id: string
+  name: string
+  description?: string
+}
+```
+
+### 4. Create the dataloader instant
+
+```ts
+const dataLoader = new DataLoader<PackageName, PackageInfo>(
+  async (/* 1 */keys: readonly PackageName[]) => { /* 2 */
+    return await Promise.all(
+      keys.map((packageName) => {
+        getData(packageName).catch((e) => e);
+      })
+    );
+  },{
+    cacheMap: new LRUMap(100/* 3 */);
+  }
+);
+```
+
+As you can see, we give the keys that the user asks for to the dataloader (1).
+
+I recommend setting the keys to be some kind of id for the following reason:  
+Lets say that in your api, the user asks for data using its id, so the user will fetch with `id:"something"`, you could just pass the id as the key, instead of changing it.
+
+The dataloader will first check if it has the keys on cache, and if so, it would return the data, without going through the database.
+
+After that (2), we give the dataloader a function for fetching the data, in case it doesn't have it in the cache.  
+In this case, I have the function `getData`, and I'm using the keys to "get data" from my database.
+
+In the end (3) we give it cacheMap and some value, the value represents how many queries dataloader will cache, in this case, after a 100 values, it will delete the first value to make space for the 101th value.
+
+From now, to query data, you just run
+
+```ts
+dataLoader.load(keys)
+```
+
+# Dataloader with GraphQL
+
+Dataloader was designed to work with GraphQL, to solve the [n+1 problem](https://medium.com/the-marcy-lab-school/what-is-the-n-1-problem-in-graphql-dd4921cb3c1a)
+
+## How to use the dataloader package with GraphQL?
+
+To use the dataloader with GraphQL just pass it in the context!
+
+```ts
+type PackageName = string;
+type PackageInfo = Package;
+export type GraphQLContext = {
+  dataLoader: DataLoader<PackageName, PackageInfo>;
+  ...
+};
+const dataLoader = new DataLoader<PackageName, PackageInfo>(
+  async (keys: readonly PackageName[]) => {
+    return await Promise.all(
+      keys.map((packageName) => {
+        npmApi.getPackage(packageName).catch((e) => e)
+      })
+    );
+  },
+  {
+    cacheMap: new LRUMap(100),
+  }
+);
+...
+export async function contextFactory() {
+  return { dataLoader, ... };
+}
+```
+
+Now, in your resolvers, just call `context.dataLoader.load(keys)` and that's it! You now have caching in your server!
```

Comment:
Do we have other articles or video links for Further Reading?

Reply:
Not that I know of, but I'll search for some